server/src/main/java/se/fortnox/reactivewizard/server/RxNettyServer.java

```diff
@@ -65,17 +75,50 @@ public void run() {
     }
 
     private void registerShutdownHook() {
-        Runtime.getRuntime().addShutdownHook(new Thread(() -> shutdownHook(config, server, connectionCounter)));
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> shutdownHook(config, server, eventLoopGroup, connectionCounter)));
+    }
+
+    public static void registerShutdownDependency(Runnable blockShutdownUntil) {
+        if (RxNettyServer.blockShutdownUntil != null && blockShutdownUntil != null) {
+            throw new IllegalStateException("Shutdown dependency is already registered");
+        }
+        RxNettyServer.blockShutdownUntil = blockShutdownUntil;
     }
 
-    static void shutdownHook(ServerConfig config, HttpServer server, ConnectionCounter connectionCounter) {
-        LOG.info("Shutdown requested. Will wait up to {} seconds...", config.getShutdownTimeoutSeconds());
-        server.shutdown();
+    static void shutdownHook(ServerConfig config, HttpServer server, EventLoopGroup eventLoopGroup, ConnectionCounter connectionCounter) {
+        LOG.info("Shutdown requested.");
+        awaitShutdownDependency();
+        LOG.info("Will wait up to {} seconds...", config.getShutdownTimeoutSeconds());
```

Comment:
The log should probably be before the above row, and you should only allow waiting for the configured time.